File: code/iFrALM.py
__all__ = ["iFrALM"]

import numpy as np
from math import *
from fralm import fralm

def ifralm(M, fixed_rank = 1, lam = 0.003, magnitude = 1e-6, batch_size = 5) :
    shape = M.shape
    epoch_num = int(ceil(shape[1] / batch_size))
    if epoch_num < 1 :
        A, E = fralm(M, fixed_rank, lam, magnitude)
        return A, E
    else :
        A_last_batch, E_last_batch = fralm(M[:, :batch_size], fixed_rank, lam, magnitude)
    print("position 1: code went here")
    U_last_batch, s_last_batch, V_last_batch = np.linalg.svd(A_last_batch, full_matrices = True)
    n = batch_size
    batch_inc_order = 1

    for i in range(1, epoch + 1) :
        if n + batch_size > shape[1] :
            l = shape[1] - n
        else :
            l = batch_size
        D = M[:, n : n + l]
        sgn = np.sign(D)
        scl = np.maximum(np.linalg.norm(D, 2), np.linalg.norm(D, np.inf) / lam)
        
        A = np.zeros((shape[0], l))
        E = np.zeros((shape[0], l))
        Y = sgn / scl
        A_last_iter = A
        E_last_iter = E
    
        # parameters from the paper
        s = np.linalg.svd(Y, full_matrices = False, compute_uv = False)
        mu = 0.5 / s[0]
        pho = 6

        in_stop = magnitude * np.linalg.norm(D, 'fro')
        out_stop = in_stop / 10

        in_converge = False
        out_converge = False
        
        #iFrALM
        epoch = 0
        print("position 2: code went here")
        while not out_converge:
            while not in_converge:
                print("position 3: code went here")
                U, s, V = isvd(U_last_batch, s_last_batch, V_last_batch, D)
                print("position 4: code went here")
                A = np.dot(U * s, V[:, n : n + l])
                E = soft_threshhold(D - A + Y / mu, lam / mu)
                in_converge = np.linalg.norm(A - A_last_iter, 'fro') < in_stop and np.linalg.norm(E - E_last_iter, 'fro') < in_stop
                # print("position 3: code went here")
                print("A convergence: " + repr(np.linalg.norm(A - A_last_iter, 'fro')))
                print("E convergence: " + repr(np.linalg.norm(E - E_last_iter, 'fro')))
                A_last_iter = A
                E_last_iter = E

            out_error = np.linalg.norm(D - A - E, 'fro')
            out_converge =  out_error < out_stop
            print("incremental batch order: " + repr(np.linalg.norm(D - A - E, 'fro')) + "epoch : " + repr(epoch) + "error : " + repr(out_error))

            U_tilde, s_tilde, V_tilde = np.linalg.svd(np.transpose(np.transpose(V[:, 1:n] * s)))
            U_last_batch = np.dot(U, U_tilde)
            s_last_batch = s_tilde
            V_last_batch = V_tilde
            Y = Y + mu * (D - A - E)
            mu = pho * mu
            epoch = epoch + 1

            
        U_last_batch = U
        s_last_batch = s
        V_last_batch = V
        n = n + l
    return A, E



def isvd(P_k, s_k, QH_k, D) :
	# input full version of SVD: P_k, s_K, QH_k
	# return full version of SVD: U_i, s_i, V_i
    m, n = D.shape
    k = min(m, n)
    M = np.dot((np.eye(m) - np.dot(P_k, np.transpose(P_k))), D)
    P_k_hat, R = np.linalg.qr(M, mode = 'reduced')
    B = np.zeros((len(s_k) + R.shape[0], len(s_k) + D.shape[1]))
    B[:len(s_k), :len(s_k)] = np.diag(s_k)
    B[:len(s_k), len(s_k): -1]
    B_hat1 = np.concatenate((np.diag(s_k), np.dot(np.transpose(P_k), D)), axis = 1)
    B_hat2 = np.concatenate((np.zeros((k, k)), R), axis = 1)
    B_hat = np.concatenate((B_hat1, B_hat2), axis = 0)
    
    U_k, s_hat, VH_k = np.linalg.svd(B_hat, full_matrices = False)

    U = np.dot(p.concatenate((P_k, P_k_hat)), U_k)
    s = s_hat[:k]
    V1 = np.concatenate((QH_k, np.zeros((n, len*(s_hat) - k))), axis = 1)
    V2 = np.concatenate((np.zeros((len*(s_hat) - k, n)), np.eye(len*(s_hat) - k)), axis = 1)
    V = np.dot(VH_k, np.concatenate((V1, V2), axis = 0))

    return U, s, V


def soft_threshhold(S, tau):
    return np.sign(S) * np.maximum(np.abs(S) - tau, 0)

File: code/ifralm.m
function [A, E] = ifralm(M)
shape = size(M);
fixed_rank = 2;
lam = 0.03;
magnitude = 1e-3;
batch_size = 10;
batch_num = ceil(shape(2) / batch_size);

%% fralm for initial batch
if batch_num > 1
    D = M(:, 1 : batch_size);
else
    D = M;
end
shp = size(D);
sgn = sign(D);
scl = max(norm(D, 2), norm(D, inf) / lam);
	
A = zeros(shp);
E = zeros(shp);
Y = sgn / scl;
A_last = A;
E_last = E;

s = svd(Y, 'econ');
mu = 0.5 / s(1);
pho = 2;

in_stop = magnitude * norm(D, 'fro');
out_stop = in_stop / 10;

in_converge = false;
out_converge = false;
while ~out_converge
    while ~in_converge
        disp('We are here 1')
        [U, S, V] = svd(D - E + Y / mu, 'econ');
        S = soft_threshhold(S, 1 / mu);
        if sum(diag(S)) < fixed_rank 
            A = (U .* diag(S)') * V';
        else
            vec = diag(S)';
            vec(fixed_rank + 1 : end) = 0;
            A = (U .* vec) * V';
        end
        E = soft_threshhold(D - A + Y / mu, lam / mu);
        in_converge = norm(A - A_last, 'fro') < in_stop && norm(E - E_last, 'fro') < in_stop;
        disp(strcat('A convergence :', num2str(norm(A - A_last, 'fro'))))
        disp(strcat('E convergence :', num2str(norm(E - E_last, 'fro'))))
        A_last = A;
        E_last = E;
    end
    in_converge = false;
    out_converge = norm(D - A - E, 'fro') < out_stop;
    disp(strcat('D convergence :', num2str(norm(D - A - E, 'fro'))))
    Y = Y + mu * (D - A - E);
    mu = pho * mu;
end
if batch_num == 1
    return
end
%% iFrALM
U_lastbatch = U(:, 1 : fixed_rank);
S_lastbatch = S(1 : 1 : fixed_rank, 1 : 1 : fixed_rank);
V_lastbatch = V(:, 1 : 1 : fixed_rank);

batch_sumlen = batch_size;
batch_order = 2;
disp('we are here 2')
for i = 2 : batch_num
    if batch_sumlen + batch_size > shape(2)
        batch_len = shape(2) - batch_sumlen;
    else
        batch_len = batch_size;
    end
    D = M(:, batch_sumlen + 1 : batch_sumlen + batch_len);
    sgn = sign(D);
    scl = max(norm(D, 2), norm(D, inf) / lam);

    A = zeros(size(D));
    E = zeros(size(D));
    Y = sgn / scl;
    A_lastiter = A;
    E_lastiter = E;
    
    s = svd(Y, 'econ');
    mu = 0.5 / s(1);
    pho = 6;
    
    in_stop = magnitude * norm(D, 'fro');
    out_stop = in_stop / 10;

    in_converge = false;
    out_converge = false;
    
    while ~out_converge
        while ~in_converge
            [U, S, V] = isvd(U_lastbatch, S_lastbatch, V_lastbatch, D - E + Y / mu);
            A = (U .* diag(S)') * V(batch_sumlen + 1 : batch_sumlen + batch_len, :)';
            E = soft_threshhold(D - A + Y / mu, lam / mu);
            in_converge = norm(A - A_lastiter, 'fro') < in_stop && norm(E - E_lastiter, 'fro') < in_stop;
            A_lastiter = A;
            E_lastiter = E;
            disp('we are here')
        end
        in_converge = false;
        out_error = norm(D - A - E, 'fro');
        out_converge =  out_error < out_stop;
        [U_tilde, S_tilde, V_tilde] = svd(S * V(1 : batch_sumlen, :)');
        U_lastbatch = U * U_tilde;
        S_lastbatch = S_tilde;
        V_lastbatch = V_tilde;
        Y = Y + mu * (D - A - E);
        mu = pho * mu;
    end
    U_lastbatch = U;
    S_lastbatch = S;
    V_lastbatch = V;
    batch_sumlen = batch_sumlen + batch_len;
    batch_order = batch_order + 1;
end



end



%% soft threshhold
function S =  soft_threshhold(S, tau)
     S = sign(S) .* max(abs(S) - tau, 0);
end

%%iSVD
function [U, S, V] = isvd(P, S, Q, D)
    % Ak =  P * S * Q'
    r = size(S, 1);
    [P_tilde, R] = qr((1 - P * P') * D, 0);
    
    B = zeros(size(S) + size(R));
    B(1 : r, 1 : r) = S;
    B(1 : r, r + 1 : end) = P' * D;
    B(1 : size(R, 1), r + 1: end) = R;
    
    [U_tilde, S_tilde, V_tilde] = svd(B, 'econ');
    U_tilde = U_tilde(:, 1 : r);
    S_tilde = S_tilde(1 : r, 1 : r);
    V_tilde = V_tilde(:, 1 : r);
    U = zeros(size(P, 1), size(P, 2) + size(P_tilde, 2));
    U(:, 1 : r) = P;
    U(:, r + 1 : end) = P_tilde;
    U = U * U_tilde;
    S = S_tilde;
    V = zeros(size(Q) + size(R));
    V(1 : size(Q, 1), 1 : size(Q, 2)) = Q;
    V(size(Q, 1) + 1 : end, size(Q, 2) + 1 : end) = eye(size(R, 1));
    V = V * V_tilde;
end
